### DISCUSSION TITLE
Map Export Templating

### DISCUSSION THREAD

**Discussion by spencerwahl** [fgpv-vpgf/fgpv-vpgf#4113] [Team Posts]
## Current situation
We have requirements to add map-author control to our map exports. For instance having the legend vertical and beside the map instead of below and split over columns.

Right now we allow titles and footers to be customized in the config, but nothing to do with layout.

## What we want to end up with

- Ability to specify where "known" parts should go; legend, map, possibly scalebar and north arrow
- Ability to add text wherever
- Ability to add pictures? (Maybe they want to supply a simplified legend for exporting)

### 1. Markdown based solution

Augment markdown with our own 'anchor points', text and images handled by normal markdown rules.

Markdown is already used for infoSections in the legend as far as I know, the conversion we do to SVG is very basic.

E.g.
```
## Wow this export is great
![](myimage.jpg)
${legend} ${map}
Look at all those layers!
```

`${legend}` and `${map}` would be replaced by the svg's generated by our export. We would need to enhance the generation to be able to take into account the relative sizes of each portion.

PROS:
- Very flexible, map-authors can put text or images wherever
- We already use markdown 

CONS:
- Need to convert markdown output to SVG
- Adding markdown doesn't do anything to reduce complexity, may drive it up

### 2. HTML
Much like #1 we let the user define an HTML file and use that to build our export.
Again, we have to the convert the actual HTML portions to SVG, like we do for everything currently.

PROS:
- Flexible
- We already convert HTML

CONS:
- More converting again
- Have to think about supporting styles and such (what happens if the user `float: "left";`s an image

### 3. Custom 'simple' markup
Have set `anchors points` only, allow user to dictate their relative placements

| |  |
| ------------- | ------------- |
| ${title}      | < ( continue title) |
| ${legend} | ${map} |
| ^ (continue legend) | ${description} |

PROS:
- Don't have to convert one markup to another, its just a guide for where to place things

CONS:
- Less flexible
- Text blocks need to fit into a predefined anchor like `description` or `footnote`
- Need to anticipate what people want

### 4. Pick a predefined style
We create some predefined templates, map authors can specify which they choose.

E.g. 
`export-style: "classic"`
`export-style: "vertical legend"`
`export-style: "only map"`

PROS:
- Don't have to handle *every* configuration
- We can test all of our defined templates

CONS:
- Less freedom for map authors
- If someone needs something special we have to add it in later
- Again, need to anticipate what people want

## Points of discussion
- Do we anticipate needing to support anything else?
- Which if any of the suggestions seem promising to you?
- Any other ideas for this problem. (Alternatively, have I overlooked anything that makes this very easy.)
- Possible issues you see arising.

**Comment by james-rae** [comment]
### Markdown.

> map-authors can put text or images wherever

During the latest docs update, i found this is true when going down the page, but when doing side-to-side stuff it was much more fickle (e.g. could not get a nice solution to text on left, picture on right, keep as a group). 

Following that, the example posted looks like it follows the "down the page" format.  If someone wanted legend on left, map on right, I don't know how much we can support.

**Overall gut feel**: Probably the best balance of simplicty and rich-features, but I see users asking for the side-by-side support very fast.

### Html.

Solves the markdown layout with div magic.
Template author needs to be a little more tech savvy, deal with piles of tags, etc. I.e. a non programmer could get markdown working by reading a 1-page guide; html they could likely only do the basics in a short period.

**Overall gut feel**: The most flexible. The most prone to edge cases. Would be interesting to see how well an html-to-svg conversion of a mockup for what ccp wants performs.

### Custom simple

Could we take this idea of the predefined "blocks" and allow markup to be inserted in each?  Removes your one PRO but lets us leverage all the nice formatting markdown gives us.

**Overall gut feel**: I applaud the simplicity but feel people will never be satisfied with limitations.

### Predefined style.

Depending how the "styles" are encoded, would an api call to "add style" solve the inflexibility?  Though tbh the "style" might end up resembling a janky subset of html, so might be more effective to just go html option at that point.  But I'm almost thinking of something like how bootstrap css lets you define a row, and then how many columns that row is split into.

Will also apply the "insert markdown into a section" variant here.

**Overall gut feel**: I like this idea, in particular if it turns out to be relatively low efforts to add new styles internally (i.e. don't need a massive module for each); could be student task to add common patterns found by googling "web panel css layout".  If we have enough, may reduce the complaints if others can't add their own style.

**Comment by AleksueiR** [comment]
I think our best bet would be to create predefined templates and let the map author pick one.

Trying to provide tools for map authors to define the export template is going to be very hard if we want the end result to be stable and generally usable, especially if they want to move elements around, and not just reorder how the existing elements are stacked. RAMP is not a PowerPoint.

Displaying the legend to the side of the map image presents its own challenges especially if the legend is long. Do we wrap it somehow or do we truncate it or do we force map authors to pick which elements will be visible in the legend because it's too long and doesn't fit?

There should not be that many different configurations (at least I hope so) that we can meaningfully arrange what export elements we have right now into. Letting map authors insert some content through Markdown into some of these predefined slots should be enough for most cases. 

If people want some crazy fancy layout, they should just export the map image and make it themselves in PP or something. I've only seen one layout request so far - legend beside the map - and don't think we should spend enormous efforts on turning our export functionality into a graphics editor.

**Comment by james-rae** [comment]
> Displaying the legend to the side of the map image presents its own challenges especially if the legend is long. Do we wrap it somehow or do we truncate it or do we force map authors to pick which elements will be visible in the legend because it's too long and doesn't fit?

Thats a good point. A compromise might be a stated limitation; published maps with structured legend and a civilized legend can use this to good effect.  We just make it clear that massive legends will render like garbage.  

Anyway just one way.

**Comment by spencerwahl** [comment]
The only thing I was really worried about for the predefined templates was complexity in maintenance. Having to make a new template for each project we use RAMP in is a less than ideal solution. Allowing markdown in certain sections would hopefully reduce that need. As @AleksueiR said, how many ways can you really orient the map and legend relative to eachother.

Also, even choosing the predefined templates we'll get more knowledge in manipulating our export... and its possible we stumble upon some easy abstraction method.

>Displaying the legend to the side of the map image presents its own challenges especially if the legend is long. Do we wrap it somehow or do we truncate it or do we force map authors to pick which elements will be visible in the legend because it's too long and doesn't fit?

Predefined templates allow us extra control over behaviour, but I think as long as we'd have samples for map-authors to see it'd be fine (i.e. show which of: push text to the side, leave empty space beside super long legend, etc. would happen)

**Comment by mweech** [comment]
Here are my thoughts based on the above comments:

- Providing a set of "default" layouts to choose from would make sense.  Similar to how most blog editors would have a typical layout for a blog, we would do the same for maps.  
- Providing limited formatting ability within the template sections makes senses.  It provides some basic formatting abilities while working within an existing layout.  Is Markdown really a good option though?  Wouldn't James example of a limitation with Markdown apply?  Should we pursue HTML with some guidance on "don't do  dumb things x, y & z"? 
- My only concern with the strategy is what is the method to deliver a new/different layout??  Would we need to make a new RAMP release?  Is there some way where we could disconnect the need to make a new build of RAMP from the ability to define a basic layout?  i.e. should we be able to add a map template to a "templates folder"?  This could be something that is added layer but we should at least think about this scenario.
- Agree on showing samples.  Similar to how the blog templates showed a sample, we would need to do the same for the layout templates.

**Comment by AleksueiR** [comment]
# Pontification

## tl;dr

Export layouts are predefined; a map author can pre-select which export layout will be used through the config with an option to allow the user to change the layout. An export image with any layout can be generated through an API call.

## Predefined layouts

Right now, @spencerwahl and I are thinking the best approach in terms of efforts and payoff is to provide a set of predefined export layouts _without_ giving map authors ability to create their own layouts. Arguments as follows:

- there is only a limited set of way map, legend and some text/markdown can be arranged on the page;
- right now, we only have a single request for a custom export layout (see below);
- given a limited number of possible layouts, a new layout would likely mean making changes to the RAMP code and hence a new release;
- we suspect it will require a lot of effort to provide a way for map authors to create custom export layouts which produce decent results (a drag-and-drop editor with scale support for example would be idea way for map authors to create new layouts; using Markdown or HTML to create layout is a clunky option and can be easily abused); there needs to be a very strong case to spend time implementing this functionality;

### Layout Types

#### Stack

The existing export layout stacks all elements and the only obvious modification is to swap positions of legend and map. Text blocks can be added in between other elements to cover all the options. This layout works well and produces decent results with any legend, so we think this is most suitable for RAMP instances without a predefined config (like catalog maps for example).

Here's a sample (original on the left; modification on the right):

![](https://i.imgur.com/lc5DLnb.png)

### Side-by-side

CCCS requested a layout where the legend is placed besides the map image (on the left or right). With this layout, only two text blocks can be used - above or below the map. This layout will not produce good images if the legend is very long as it will create a lot of white space underneath the map. It would best used for relatively short legend - maps with predefined configs and a few layers.

Here's a sample:

![](https://i.imgur.com/xrNgdHw.png)

### Overlay

This is a variation on the side-by-side arrangement of elements where the legend is placed directly on top of the map image. There are only two text blocks as well. This layout can only work with very small legends (just a couple of layers without any sublayers) because the legend obscures part of the map. The legend can be placed in one of the corners of the map image.

Here's a sample:

![](https://i.imgur.com/MOsfjce.png)



## Layout Modes

CCCS requests a custom export layout (where legend can be placed to the side of the map image) so the resulting image can be inserted into a report. Let's call this **Report Mode**.

Layout Modes are separate from the layout types, and both modes should support all the possible arrangements.

### Classic Mode

This is the layout used right now. We do not maintain any aspect ratio and don't try to scale font sizes to fit into a regular page. In this mode, the export image can be ridiculously long if you just keep adding layers to the map.

### Report Mode

It seems that a crucial point in the CCCS request is being able to insert the export image into a report - the image should fit into a standard page (assume 11 x 8.5) and any text on the image should remain legible after scaling (need to think about font sizes here). From this, we also get Portrait and Landscape variants of the layout.

In the report mode, there will be no option to add blocks of text to the image because that can change the size/ratio of the resulting image. A single line for a title at the top or a caption at the bottom of the image should be possible. Since this image is generated to be included in a report, any additional text information should come direction from the surrounding text.

The last point needs to be confirmed. Would a single caption/title line be enough for CCCS needs or not?

#### Landscape

In landscape, we further assume that an export image would *always* fill the full page and therefore has a fixed 11 x 8.5 ratio and correspondingly sized text. 

Here are samples of landscape export images (no legend, legend at the bottom, legend on the side, floating legend):

![](https://i.imgur.com/DJSgdSe.png)

#### Portrait

In Portrait, the export image can either take the full page (fixed ratio) or be inserted in the middle of the document.

![](https://i.imgur.com/hslhSeN.png)

Technically, the full page Portrait layout is a special case of an export image intended to be inserted in the middle of a page. The ratio of this export image should not be smaller than 8.5 x 11. We also need to make sure that when the image is scaled to fix the width of the document, the fonts are readable.

In the following example, text above and below the export image belongs to the report document.

![](https://i.imgur.com/YfwjjZW.png)

## Config and API

A specific layout (layout mode, layout type, and any text which can be added to the layout) can be set in the config as the default with a possible option to allow the user to change the layout later.

If we want to allow users to switch layouts, we would need to design a UI component for that. It will probably include a two-step process: selection layout mode and then selection layout type, with some further options where to put the legend (above/below, left/right, one of four corners when floating).

RAMP API should allow to generate any export image of any layout regardless of the config settings.

When selecting the report mode, the map author should test and ensure that the legend looks good in that layout - long legend will overflow the image and be truncated.

## Good to have

- a config option to specify which layers and layer entries should be included in the export legend
- a UI component that lets users to specify which layer entries should be included in the export legend

## Discussion points

- Can you think of any more ways (which make sense) to arrange export elements?
- Are there any other good arguments for allowing config authors to create their own export templates?
- Do you think splitting exports into Report and Classic mode makes sense?
- Do we want to allow users to switch export layouts at runtime (might be useful to users)?
- Do we want a muffin dispenser installed in the office?
- Do we want to provide functionality to select which layer entities are included in the export legend (useful to users and config authors when making sure the select layout looks good with the legend)?
- Did I make many typos?

**Comment by spencerwahl** [comment]
> - Do we want to allow users to switch export layouts at runtime (might be useful to users)?
> - Do we want to provide functionality to select which layer entities are included in the export legend (useful to users and config authors when making sure the select layout looks good with the legend)?

These may be good as future enhancements rather than initial requirements. I think we can get by without them at first (or maybe entirely).

> - Do we want a muffin dispenser installed in the office?

We must have it dispense snacks for all dietary restrictions in the name of respect.

**Comment by mweech** [comment]
> Would a single caption/title line be enough for CCCS needs or not?

Based on what they described so far.. no.  I am going to circle back and ask for some samples on what they have in mind for a map export layout so we can parse this some more.  However, what they described is the ability to have a map exportable that could be used in a report (i.e. doc, or ppt presentation) and so would have a condensed layout such as a legend at the side of the map.  They also want to be able to change the text that would appear in the export, in terms of a title/description/footnote that would be more personalized to their application.   i.e. all climate normals maps export an image with a disclaimer/remark re: Climate Normals vs a DCS export where they would make a disclaimer/remark about that specific dataset (i.e. compared to reference period yadda yadda).

Other thoughts for now:
- ability to define what layers actually show up in the export legend is important.  i.e. CCCS have no interest in addng contextual layers to the print export, but the climate layers should be described.
- rather than allowing the layout to be "modified" why not just have left sided vs right sided legends..?  or map first stack vs legend first stack
- do we really need a layout mode..? isn't this a bit of skirting around functionality that boils down to "allow map author to define size and orientation of map image"? layout would flow around the map image...

**Comment by ShrutiVellanki** [comment]
### **WRT** text skewing export image sizes: 

Would it be possible to make use of the `<figcaption>` HTML tag (or similar) here?  I'm thinking they won't need more than a paragraph's worth of a caption + a couple sentences of a footnote so: 
- Set a maximum allowed area that the `<figcaption>` tag takes up
- Allocate a certain (small, i.e. couple sentences) amount of the maximum area to a footnote (that they can toggle on off)
- Allocate the rest  (larger, i.e. small paragraph) of the area to any caption they might want to put (they can toggle this on/off)
- Truncate any text that doesn't fit within the area with `...` so they know they've exceeded caption limits 
- All text that falls into the caption would be smaller than the traditional export image text with 
- Footnote font size could be smaller than caption size

I'm thinking it would end up looking something like this: 
![](https://cpb-us-w2.wpmucdn.com/edblogs.columbia.edu/dist/6/789/files/2015/09/Slide3-1hjo2zh.jpg)

As you can see the figure (in our case the map) size still controls the export image ratios. [More examples.](https://edblogs.columbia.edu/eescx3050-001-2015-3/scientific-figures/)

With a predefined max area, text won't skew the export image ratios past `100px - 300px` vertically and wouldn't skew it horizontally. With toggling options on off we could make it so that they can only export image + legend if they so choose, while also ensuring that they don't try to cram a whole report's worth of text into the export image.

**Comment by AleksueiR** [comment]
> do we really need a layout mode..? isn't this a bit of skirting around functionality that boils down to "allow map author to define size and orientation of map image"? layout would flow around the map image...
@mweech

That would be a good solution, but it's not possible to set the absolute size of the map export image because it ultimately depends on the size of your screen. With local-JS exporting, the export image of the map itself is always the same as the RAMP's container on the page. It's not possible to scale up the map image using local-JS logic without losing quality (with the previous print service approach it was possible to set the output size). We even had to remove the option to select the size of the export image we had before. See here: https://github.com/fgpv-vpgf/fgpv-vpgf/issues/3044#issuecomment-425940220

> > It might be possible by changing the size of the map itself, waiting for everything to load up, rendering the resulting map on the canvas, and resizing the map back.

The above sounds like it could work, but we haven't tried it and are not sure how reliable/fast this approach would be. If this is possible, then yes, it would be possible to specific the absolute export image size.

>  I am going to circle back and ask for some samples on what they have in mind for a map export layout so we can parse this some more. 
@mweech

Absolutely. We need to know more of what they want. My assumptions are based on that single CCCS change request. I know they said they want to change text on the export image itself, but I still think that if that image to be inserted into a Word document, they don't need more than maybe a couple lines of text on the image. All supporting text, in my opinion, should be put in the document itself. 
This is exactly what @ShrutiVellanki says - provide a fixed amount of space for the caption/footnote (toggleable) and truncate anything that doesn't fit. We would have to play with font sizes in this case as well because the full size of the export image cannot be predicted right now.

> rather than allowing the layout to be "modified" why not just have left sided vs right sided legends..? or map first stack vs legend first stack
@mweech

You are talking about the UI component for the use to select the layout, right? Yes, this could be the way to do it. I was thinking of providing small images of different export options for the user to chose from, not just a dropdown with text labels.

> ability to define what layers actually show up in the export legend is important. i.e. CCCS have no interest in adding contextual layers to the print export, but the climate layers should be described.
@mweech

Fair enough. Initially, we can implement this internally through the config and API, and develop the UI later.

**Comment by mweech** [comment]
Brief chat with @james-rae and @spencerwahl seperately about this today.  It strikes me that this is getting very complex to add **agility and flexibility** to export map when 1 client is really looking for **specificity** in their own exports.    I think we have to take a step back and say does all this effort justify the pay off... I suspect that might be: NO (for now). 

So question: should we instead look at the ability to have "export map" plugins...? This would allow a project to replace the default "Best effort" output with a defined layout that suits their application requirements.  For instance, for CCCS they would only export the map and specific legend in the layout they desire and are free to add in the specific text labels they want for each export (i.e. the time period selection, etc).   Our own export functionality would continue to use the existing layout code. ..?

**Comment by AleksueiR** [comment]
This is a very good point. If we create an export plugin specifically for a certain application's requirements, we can avoid complexity of trying to deal with the multitude of edge cases and possible layouts and options.

**Comment by milespetrov** [comment]
I guess this depends on if map authors have the knowledge/resources and time to build a plugin vs. a more basic templated option. There would be some additional work needed to make exporting a replaceable feature in ramp. A plugin would work nicely if there is outside interest to create these types of plugins. 

I'm guessing the plugin can ask ramp for each piece of an export (map, legend etc) then it smashes it together and gives ramp back a completed image for the user to download/view. We'd need to plan/design how these interactions would take place (via the API, or special methods for this type of plugin) and how complex this would be for a plugin author to implement on their own (is smashing and styling images together easy-ish, should we be concerned with this?).

**Comment by spencerwahl** [comment]
# Pontification part 2

All of this comes from a pod discussion with @AleksueiR 

### tl;dr
- Generators accessible by plugins, called to get chunks of export back such as legend or map
- Plugins give RAMP a set of SVGs and x,y positions, RAMP smashes images back together
- Plugins either replace export like how the table works or we use a combo of disabling normal export and an export event


## Starting the export

### 1. Export-replacing plugins

Like with the table we allow plugins to take over the 'export slot'. We would then require some sort of `generateExport` function that would be auto-called on the plugin. RAMP calls that function when the user clicks for an export and we let the plugin handle the rest.

### 2. Regular plugins

For this case we would need either an observable or an event to let the plugin know its export time. We would also need a way to turn off the regular export functionality (like what we do with different identify modes).

## Generating the chunks

We will allow access to the generator functions we currently use. This includes the Map, Legend, North arrow, scalebar, and footnote (text). Export plugins can call these generators with parameters (partially explained below) and get back an SVG.

### Legend generator

The legend generation is the most complicated section of the current export. In order to utilize what we currently do we will let plugins specify a number of columns and or width they want, and (assuming feasibility) a subset of the legend that they want to include.

e.g. To get a 1 column legend with only 2 layers out of 4
```
// (legend, columns, width in px)
api.legendGenerator([layer2, layer4], 1, 200);
```

### Map generator

The Map generator will function completely as it does in app (for client-side printing). We will return the 1:1 sized map image. Plugins can scale up or scale down the SVG as desired.

### North arrow and scalebar

Function as they normally do

### Footnote/text

Right now this generator handles only text (as far as we can tell). We think that the [HTML to Canvas library](https://html2canvas.hertzen.com/) should be able to handle any simple HTML. Plugins can either import that library themselves or we can use it internally if useful. If we decide to expose this generator then it can handle text blurbs for plugins.

### Extra chunks

Plugins can generate other chunks they want on their own (other images, etc.). As long as they end up as an SVG we can add them to the final export.

## The SVG collage

There are 2 big options that we see:

### 1. Make plugins do it

Not much explanation here, plugins smash together all the SVGs and give that back to RAMP. RAMP displays in export and we are done ✅ 

### 2. RAMP enhance (aka we don't want to rewrite the smash code every time)

Rather than an endpoint that just accepts an image, we provide an endpoint that takes an array of image and position pairs (and probably a canvas size).

`[[legendSVG, (0, 200)], [mapSVG, (100,200)]]` etc.

RAMP will then take this array and blindly mash together the SVGs on to the canvas at the specified position. The final mashed up image is then displayed in export.

In this case plugins are still 100% responsible for determining the position and size of each bit of the export, as well as providing any extra chunks that RAMP normally wouldn't have. What this does provide is eliminating the need to duplicate the SVG mashing code, it can also reuse the existing stack if no positions are provided. Of course, if this becomes more work than expected we can offload everything to plugins as in option 1.

## Discussion points

- Export replacing plugins vs regular
- Issues with generators
- Is it too much work to do the RAMP enhancement

**Comment by mweech** [comment]
Thoughts...
-Export replacing plugins sounds good!
-re: generators.. a footnote is just text... since adding text with some formatting is pretty common use case, i suggest not having a "footnote" or plain text generator and instead let the plugin make their pretty SVG of text that just gets added into the smashing of images together.  maybe this becomes better after we do the first custom export or two... 
-RAMP enhance sounds pretty flexible!

**Comment by AleksueiR** [comment]
Just wanted to clarify that export plugins will not replace the export dialog itself right now, only the logic generating the final image. Conceivably, someone can build a plugin with its own export dialog and implement that drag-n-drop editor to scale/move export elements around.

I agree with @mweech - we should just have a **text** (well, technically **HTML**) generator instead of having several text-like generators.

I don't think it's a good idea to force plugins to create the final image:
- this is duplication of smashing code;
- plugins would need to also duplicate code that checks for tainted canvas pieces and display error messages according to the config flag;
- if the plugin authors do want to smash images in the plugin, they can still do that and simply return a single resulting image to RAMP ;

**Comment by mweech** [comment]
^ Sounds like we are coming to a conclusion on how this would work.  👏 

Should specify list order = top-down order of the final image.  Overlap at your own risk..?
> [[legendSVG, (0, 200)], [mapSVG, (100,200)]] etc.

**Comment by AleksueiR** [comment]
I think the order of the components should be bottom-up - it's easier to figure out what overlaps what, I think. Overlapping of components could be intentional, like when you want your two-layer legend to be displayed on top of the map image, or if you want the scale bar/north arrow to be rendered on the map.

**Comment by mweech** [comment]
Conversation closing... @AleksueiR to compile implementation plan and acceptance criteria here: https://github.com/fgpv-vpgf/fgpv-vpgf/issues/3563

**Comment by AleksueiR** [comment]
Added the overall implementation plan to fgpv-vpgf/fgpv-vpgf#3563.
